### ISSUE TITLE
Report gives  Invalid JSON response when adding extra columns to the entry and repeat tables

### ISSUE THREAD

**Issue by sjbirve** [meeting-room-booking-system/mrbs-code#3682]
**Describe the problem you're facing/question you have**
Hi
We are running mrbs 1.11.5 with a few extra columns to the entry and repeat tables.
After adding the extra columns to the entry and repeat tables the report function no longer works.
Summary is working fine and the extra columns shows up perfectly.
We have followed the instructions for adding columns from mrbs documentation.
https://mrbs.sourceforge.io/view_text.php?section=Documentation&file=INSTALL

See screenshots from database and frontend.

**Browser details (please complete the following information):**
 - OS: [all]
 - Browser [all]
 - Version [all]

**Server details (please complete the following information):**
 - MRBS version [1.11.5]
 - Web server: [Apache version 2.4.59  ]
 - OS: [Debian 10]
 - PHP version: [8.2.19]
 - Database used: [10.3.39-MariaDB]

**Additional context**
<img width="496" alt="upsc-mrbs-entry" src="https://github.com/meeting-room-booking-system/mrbs-code/assets/172034753/154f0a86-c134-4349-ad00-f7074574c5cc">
<img width="693" alt="upsc-mrbs_rep" src="https://github.com/meeting-room-booking-system/mrbs-code/assets/172034753/a1f69c39-477a-4b49-8d85-3809a00a5f80">
<img width="406" alt="frontend" src="https://github.com/meeting-room-booking-system/mrbs-code/assets/172034753/b3b7734d-7488-4b53-aaf4-0e8c4b3fd2c2">

**Comment by campbell-m** [comment]
Can you have a look in the Network tab of your browser's Developer Tools and see what the response is?  I suspect it's got a PHP error message in it.  And/or look in your PHP error log for the messae.

**Comment by sjbirve** [comment]
I got a  long error message in the Apache error log but i don't know if its related to the error described. We are using Joomla authentication so i trimmed the Joomla token etc.

[Mon Jun 10 09:47:32.094184 2024] [php:notice] [pid 15374] [client xxx.xx.xxx.xxx:28665] Uncaught exception 'TypeError' in /var/www/html/mrbs-1.11.5_growthcabinets/functions.inc at line 83\nMRBS\\strcasecmp_locale(): Argument #2 ($string2) must be of type string, null given, called in /var/www/html/mrbs-1.11.5_growthcabinets/lib/MRBS/Auth/Auth.php on line 335\n#0 /var/www/html/mrbs-1.11.5_growthcabinets/lib/MRBS/Auth/Auth.php(335): MRBS\\strcasecmp_locale()\n#1 [internal function]: MRBS\\Auth\\Auth::compareUsers()\n#2 /var/www/html/mrbs-1.11.5_growthcabinets/lib/MRBS/Auth/Auth.php(350): usort()\n#3 /var/www/html/mrbs-1.11.5_growthcabinets/lib/MRBS/Auth/AuthJoomla.php(135): MRBS\\Auth\\Auth::sortUsers()\n#4 /var/www/html/mrbs-1.11.5_growthcabinets/lib/MRBS/Auth/Auth.php(118): MRBS\\Auth\\AuthJoomla->getUsernames()\n#5 /var/www/html/mrbs-1.11.5_growthcabinets/functions.inc(514): MRBS\\Auth\\Auth->getDisplayName()\n#6 /var/www/html/mrbs-1.11.5_growthcabinets/report.php(934): MRBS\\get_compound_name()\n#7 /var/www/html/mrbs-1.11.5_growthcabinets/report.php(1943): MRBS\\report_row()\n#8 {main}\n\nMRBS GET: Array\n(\n)\nMRBS POST: Array\n(\n    [MAX_FILE_SIZE] => 2097152\n    [csrf_token] => 9fed5ae145f2380fcdaf645d1cdc2b78f49bbb4a9e4d11843fd9022e0e318001\n    [phase] => 2\n    [from_date] => 2024-06-10\n    [to_date] => 2024-08-09\n    [areamatch] => \n    [roommatch] => \n    [namematch] => \n    [descrmatch] => \n    [creatormatch] => \n    [match_approved] => 2\n    [match_pi] => \n    [match_gc_sday] => \n    [match_gc_lday] => \n    [match_gc_light] => \n    [match_gc_temp] => \n    [match_gc_hum] => \n    [match_gc_material] => \n    [output] => 0\n    [output_format] => 0\n    [sortby] => r\n    [sumby] => p\n    [datatable] => 1\n)\nMRBS SESSION: Array\n(\n    


Is this making any sense for you?

/Simon

**Comment by sjbirve** [comment]
Hi
There's no user with no username in Joomla. Its not possible to create a user with no username in Joomla, the system will prevent it to happen.

The only instance where no user or username is present is in the local mrbs database but since we use Joomla authentication we have never had a local user in mrbs database.
Te report function where working in previous versions of MRBS with Joomla authentication and far as i remember maybe at version mrbs-1.8.0 or mrbs-1.10.0.

/Simon

**Comment by sjbirve** [comment]
Hi
We are running Joomla version 4.4.5.

/Simon

**Comment by campbell-m** [comment]
Can you try running the attached version of lib/MRBS/Auth/AuthJoomla.php (you'll need to unzip it first) and let me know what you get in your error log?

[AuthJoomla.zip](https://github.com/user-attachments/files/15767421/AuthJoomla.zip)

You should see something like

~~~text
[11-Jun-2024 01:03:19 Pacific/Auckland] DEBUG: groups array (
  0 => 6,
  1 => 2,
  2 => 8,
)
[11-Jun-2024 01:03:19 Pacific/Auckland] DEBUG: user_ids array (
  0 => 5,
  1 => 4,
)
~~~

**Comment by sjbirve** [comment]
Hi
Two things. First of all the report works without the error when i put the file you sent, lib/MRBS/Auth/AuthJoomla.php .
This is what i get in the error log.

[Mon Jun 10 15:41:13.296750 2024] [php:notice] [pid 33536] [client 172.18.143.133:22412] DEBUG: Empty: 259, referer: https://beta.upsc.se/mrbs-1.11.5-beta/report.php
[Mon Jun 10 15:41:13.293127 2024] [php:notice] [pid 33536] [client 172.18.143.133:22412] DEBUG: user_ids array (\n  0 => 328,\n  1 => 62,\n  2 => 249,\n  3 => 555,\n  4 => 758,\n  5 => 759,\n  6 => 126,\n  7 => 259,\n  9 => 383,\n  10 => 456,\n  11 => 518,\n  12 => 546,\n  13 => 571,\n  14 => 585,\n  15 => 623,\n  16 => 639,\n  17 => 661,\n  18 => 667,\n  19 => 669,\n  20 => 686,\n  21 => 689,\n  22 => 733,\n  23 => 734,\n  24 => 745,\n  25 => 746,\n  26 => 747,\n  27 => 751,\n  28 => 777,\n  29 => 787,\n  30 => 798,\n  31 => 801,\n  32 => 804,\n  33 => 832,\n  34 => 833,\n  35 => 835,\n  36 => 836,\n  37 => 853,\n  38 => 862,\n  39 => 863,\n  40 => 867,\n  41 => 873,\n  42 => 880,\n  43 => 899,\n  44 => 902,\n  45 => 905,\n  46 => 906,\n  47 => 918,\n  48 => 923,\n  49 => 940,\n  50 => 942,\n  51 => 978,\n  52 => 989,\n  53 => 992,\n  54 => 997,\n  55 => 999,\n  56 => 1021,\n  57 => 1047,\n  58 => 1059,\n), referer: https://beta.upsc.se/mrbs-1.11.5-beta/report.php
[Mon Jun 10 15:41:13.292736 2024] [php:notice] [pid 33536] [client 172.18.143.133:22412] DEBUG: groups array (\n  0 => 31,\n  1 => 8,\n  2 => 3,\n), referer: https://beta.upsc.se/mrbs-1.11.5-beta/report.php
[Mon Jun 10 15:40:34.816817 2024] [php:notice] [pid 33219] [client 172.18.143.133:1228] DEBUG: Empty: 259, referer: https://beta.upsc.se/mrbs-1.11.5-beta/report.php


/Simon

**Comment by campbell-m** [comment]
Good, thanks.  Yes, I'd expect the report to work.  Is there anything unusual about the Joomla user with id 259?  They are appearing in the resuts of a call to `JAccess::getUsersByGroup()` but it seems the username for the user with that id is null.

**Comment by sjbirve** [comment]
Hi
Strangely enough, there is no user 259 in Joomla,. I have also looked in the database with phpmyadmin and no user with id 259 is present in the Joomla user table.

But i found a foreign key with ID 259 in the user_usergroup_map table of Joomla. I deleted that key and the error is gone.
It seems like we had a user ID 259 with null username in the  user_usergroup_map table that must have been left in some upgrade or delete action and that caused the error in MRBS.

/Simon

**Comment by campbell-m** [comment]
I've now committed a fix for this issue in [547b72f](https://github.com/meeting-room-booking-system/mrbs-code/commit/547b72fa270b4eaa707866f61f05189328bd7064).  You would still get the "Invalid JSON response" if PHP errors are displayed in the browser, but at least the error message would be clearer.

I suspect this problem has been around on your system for a while, but only manifested itself in MRBS 1.11.5 which introduced stricter type checking.  As you say, the user with ID 259 was being returned by `getUsersByGroup()` but that user wasn't present in the Joomla users table.  Strangely, if you call the Joomla method `getUser()` and give it a user_id that doesn't exist, the Joomla API will still return a user object, but with all the properties set to null, and that was what was happening with ID 259.  MRBS now checks that the username property is set and, if not, generates a warning.
